atd-vze/src/views/Users/UserForm.js

```diff
@@ -0,0 +1,289 @@
+    app_metadata: {
+      roles: ["readOnly"], // Default to lowest level access
+    },
+  };
+
+  const [userFormData, setUserFormData] = useState(defaultFormData);
+  const [isFormDataLoaded, setIsFormDataLoaded] = useState(false);
+  const [isSubmissionError, setIsSubmissionError] = useState(false);
+  const [isFormSubmitted, setIsFormSubmitted] = useState(false);
+
+  const roles = [
+    { id: "itSupervisor", label: "IT Supervisor" },
+    { id: "admin", label: "Admin" },
+    { id: "editor", label: "Editor" },
+    { id: "readOnly", label: "Read-only" },
+  ];
+
+  // Fetch existing user data if editing
+  useEffect(() => {
+    if (type === "Edit") {
+      const endpoint = `${process.env.REACT_APP_CR3_API_DOMAIN}/user/get_user/${id}`;
+      axios
+        .get(endpoint, {
+          headers: {
+            Authorization: `Bearer ${token}`,
+          },
+        })
+        .then(res => {
+          setUserFormData(res.data);
+          setIsFormDataLoaded(true);
+        });
+    }
+  }, [id, type, token]);
+
+  const handleTextInputChange = event => {
+    const updatedFormData = {
+      ...userFormData,
+      [event.target.id]: event.target.value,
+    };
+    setUserFormData(updatedFormData);
+  };
+
+  const handleRoleRadioInputChange = event => {
+    const field = "app_metadata";
+    const appMetadata = {
+      roles: [event.target.value],
+    };
+    const updatedFormData = { ...userFormData, [field]: appMetadata };
+
+    setUserFormData(updatedFormData);
+  };
+
+  // Remove fields not needed for edits (so req does not fail)
+  const cleanFormDataForEdit = userFormData => {
+    const editFields = ["name", "email", "app_metadata"];
+
+    // If resetting password, must include password and required connection field
+    userFormData.password !== "" &&
+      editFields.push("password") &&
+      editFields.push("connection");
+
+    const cleanedFormData = editFields.reduce((acc, field) => {
+      return { ...acc, [field]: userFormData[field] };
+    }, {});
+
+    return cleanedFormData;
+  };
+
+  const handleFormSubmit = () => {
+    if (type === "Edit") {
+      const endpoint = `${process.env.REACT_APP_CR3_API_DOMAIN}/user/update_user/${id}`;
+      const updatedFormData = cleanFormDataForEdit(userFormData);
+      axios
+        .put(endpoint, updatedFormData, {
+          headers: { Authorization: `Bearer ${token}` },
+        })
+        .then(() => {
+          setIsFormSubmitted(true);
+        })
+        .catch(() => {
+          setIsSubmissionError(true);
+        });
+    } else if (type === "Add") {
+      const endpoint = `${process.env.REACT_APP_CR3_API_DOMAIN}/user/create_user`;
+      axios
+        .post(endpoint, userFormData, {
+          headers: { Authorization: `Bearer ${token}` },
+        })
+        .then(() => {
+          setIsFormSubmitted(true);
+        })
+        .catch(() => {
+          setIsSubmissionError(true);
+        });
+    }
+  };
```

Comment:
Not sure it makes it any better or easier to read but I was trying to reduce some duplicate code...

```js
  const handleFormSubmit = () => {
    let submitForm; 
    const headers = {
      headers: {
        Authorization: `Bearer ${token}`,
      },
    };

    if (type === "Edit") {
      let endpoint = `${process.env.REACT_APP_CR3_API_DOMAIN}/user/update_user/${id}`;
      const updatedFormData = cleanFormDataForEdit(userFormData);
      submitForm = axios.put(endpoint, updatedFormData, headers);
    } else if (type === "Add") {
      let endpoint = `${process.env.REACT_APP_CR3_API_DOMAIN}/user/create_user`;
      submitForm = axios.post(endpoint, userFormData, headers);
    }

    submitForm
      .then(() => {
        setIsFormSubmitted(true);
      })
      .catch(() => {
        setIsSubmissionError(true);
      });
  };
```

Reply:
Nice - I like this a lot. I'm going to update this!